feat(send_queue): report progress for media uploads

Johennes opened the pull request:
This makes it possible to listen to the media upload progress when using the send queue. The progress is communicated via `EventSendState::NotSentYet`.

I haven't yet fixed / enhanced the tests or added a changelog because I'm not 100% sure about this approach and would like to get some preliminary feedback first.

- [ ] Public API changes documented in changelogs (optional)



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210187425391164

---

Edit by @Hywan:

- Address https://github.com/element-hq/element-x-ios/issues/4055

Johennes commented on bindings/matrix-sdk-ffi/src/timeline/mod.rs:
```diff
@@ -274,6 +274,46 @@ impl TryInto<Reply> for ReplyParameters {
+pub enum EventSendProgress {
+    /// A media is being uploaded.
+    MediaUploadProgress {
+        /// The index of the media within the transaction. A file and its
+        /// thumbnail share the same index.
+        index: u64,
```
This should maybe be guarded with `#[cfg(feature = "unstable-msc4274")]` because it'll always be 0 for normal (non-gallery) uploads?

jplatte replied:
Doesn't seem bad to me.

jplatte commented on crates/matrix-sdk-ui/src/timeline/event_item/local.rs:
```diff
@@ -84,3 +87,29 @@ pub enum EventSendState {
+        progress: TransmissionProgress,
+    },
+
+    /// A media has finished uploading.
+    UploadedMedia {
+        /// The index of the media within the transaction. A file and its
+        /// thumbnail share the same index.
+        index: u64,
+        /// Is the media a thumbnail?
+        is_thumbnail: bool,
+    },
```
This feels weird. Why is this a separate variant, rather than a `progress` of 0% for the next `index`, or a progress of 100% if it's the last file being uploaded?

Johennes replied:
These variants are created from `RoomSendQueueUpdate` variants and `RoomSendQueueUpdate::UploadedMedia` already existed, so I tried not to break with it. I suppose technically it wouldn't be needed anymore though because, as you say, `MediaUploadProgress` with `current == total` means it's uploaded.

jplatte replied:
Right, I'm not familiar with the send queue API. Let's wait for feedback from the team :)

Johennes replied:
It is definitely a fair question though!

Hywan commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -578,7 +579,54 @@ impl RoomSendQueue {
+                    _ => None,
+                } {
+                    let mut subscriber = progress.subscribe();
+                    let our_updates = updates.clone();
```
It's fine to rename `our_updates` to `updates` here.

Hywan commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -578,7 +579,54 @@ impl RoomSendQueue {
+                } {
+                    let mut subscriber = progress.subscribe();
+                    let our_updates = updates.clone();
+                    spawn(async move {
```
Who controls this task? If the upload is cancelled, this task should be cancelled too right? Or maybe the `subscriber.next().await` may return a `None` at some point? If that's the case, please test that behaviour specifically, and comment that out, otherwise handle this 🙂.

Johennes replied:
Yes, indeed `next()` will yield `None` once the observable (`progress`) and its clones are dropped. This happens once the final `RoomSendQueueUpdate::MediaUpload` is emited further down after the request has completed.

I've added a comment about this but could also add an explicit `abort` if that would be better.

Hywan commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -726,18 +778,25 @@ impl RoomSendQueue {
+                            room.client().media().upload(&mime, data, Some(request_config));
+                        if let Some(watcher) = progress_watcher {
+                            req = req.with_send_progress_observable(watcher);
+                        };
+                        let res = req.await?;
```
```suggestion
                        };

                        let res = req.await?;
```

Hywan commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -1849,6 +1913,29 @@ pub enum RoomSendQueueUpdate {
+        is_thumbnail: bool,
+    },
+
+    /// A media has been successfully uploaded.
```
The comment is wrong.

bnjbvr commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -726,18 +778,25 @@ impl RoomSendQueue {
+                        let mut req =
+                            room.client().media().upload(&mime, data, Some(request_config));
+                        if let Some(watcher) = progress_watcher {
+                            req = req.with_send_progress_observable(watcher);
```
If I understand correctly, this will emit the following updates, if I'm sending a file with a thumbnail:

- first progress updates for the thumbnail, going from 0 to 100%
- then progress updates for the file itself, going from 0 to 100%

Is that right? I think observers would likely be more observed in the overall progress from 0 to 100%, so it would be nice if we could aggregate those into a single stream, for observers (especially for galleries which may contain many medias and thumbnails).

Johennes replied:
I think this goes together with your [other comment](https://github.com/matrix-org/matrix-rust-sdk/pull/5008/files#r2108440747). I actually cannot think of a reason why a client would be interested in displaying the thumbnail and file upload progress separately. So aggregating them would probably be best.

I think a difficulty with this is determining the overall upload size ahead of time. The way the code is currently structured, we only receive the size of the file upload once the file has been encrypted and actually started uploading. To aggregate it, we would have to know the size already when starting the thumbnail upload, however.

Johennes replied:
Maybe we could update `UploadEncryptedFile` so that it can either take a reader for the plain data (like today) or a vector of the encrypted data? Then we could encrypt both thumbnail and media ahead of time, to compute the overall upload size, store it on the send queue requests and use it inside `RoomSendQueue::sending_task` to aggregate the progress per thumbnail / media pair. That would mingle the concerns a little of course.

The only other option I can think of is to do some heuristics and translate the bytes sent for the encrypted attachments into bytes of the unencrypted attachments. That would mean the SDK wouldn't report the actually uploaded bytes though which doesn't feel great either.

🤔

Johennes replied:
Oh, yes good point that about the memory limitations when encrypting everything upfront. 

I think the percentage should probably suffice in most cases. Maybe we could just map the internal `TransmissionProgress` to a new struct that contains only the current percentage then.

Johennes replied:
I've implemented this by introducing

```
pub struct RelativeTransmissionProgress {
    /// The completion percentage as a number between 0 and 1.
    pub percentage: f32,
}
```

which worked relatively ok apart from having to add three further fields on the request kinds to track the different file sizes.

I've realized too late that a floating point percentage might introduce subtle precision issues. Maybe it's better to track the progress as a pair of `current` / `total` integers, possibly with an `fn percentage() -> f32`?

The other shortcoming is that it's not (easily) possible anymore to track _just_ the thumbnail upload progress (or completion). That is probably not a huge use case though.

Johennes replied:
@bnjbvr do you have any thoughts on how to proceed here?

Johennes replied:
Resolving this since `AbstractProgress` has been implemented in the latest version of the PR.

bnjbvr reviewed (commented):
Interesting, and nice that we could in theory have this! I've got a few questions about the approach below. We'd like to try it out before validating the overall approach, because of the remark related to memory usage across the FFI, but other than that, it would be a great addition!

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -1849,6 +1913,29 @@ pub enum RoomSendQueueUpdate {
+    },
+
+    /// A media has been successfully uploaded.
+    MediaUploadProgress {
```
Yeah, good point. They're almost identical. Have merged them into one.

jmartinesp commented:
I just tested this on EXA and it seems like there's a memory when receiving the upload progress:

https://github.com/user-attachments/assets/7dc8a55e-4237-4c12-a10e-ed4d72c4c710

Keep in mind though that this may also be caused by the upload process itself, since I'm also seeing this in the logs:

```
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 262144, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 270336, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 278528, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 286720, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 294912, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 311296, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 319488, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 335872, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 344064, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 352256, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 360448, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 368640, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 401408, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 409600, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 425984, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 442368, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 450560, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 458752, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 466944, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 483328, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 491520, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 499712, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 516096, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 532480, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 540672, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 548864, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 557056, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 598016, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 614400, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 622592, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 647168, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 663552, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 671744, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 688128, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 704512, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 720896, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 745472, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 753664, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 761856, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 794624, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 802816, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 819200, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 966656, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 974848, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1007616, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1024000, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1056768, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1064960, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1081344, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1097728, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1122304, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1130496, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1146880, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1155072, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1171456, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1187840, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1204224, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1228800, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1236992, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1253376, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1286144, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1294336, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1318912, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1335296, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1400832, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1417216, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1425408, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1679360, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1687552, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1695744, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1703936, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1712128, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1720320, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1728512, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1744896, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1761280, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1967280, total: 1967280 } }) }}
W  matrix_sdk_ui::timeline::controller: Timeline item not found, can't update send state | crates/matrix-sdk-ui/src/timeline/controller/mod.rs:930 | spans: local_echo_handler{room_id="!ssEkHKesfRtdXQRDYP:matrix.org" focus="pinned-events" prefix="pinned_events"} > update_event_send_state{txn_id="7fea59fc56b34b48a34a027c668f5c9b" send_state=NotSentYet { progress: Some(MediaUpload { index: 0, is_thumbnail: false, progress: TransmissionProgress { current: 1967280, total: 1967280 } }) }}
```

So I'm not sure the timeline events are even being updated.

Hywan commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -578,7 +579,54 @@ impl RoomSendQueue {
+            };
+
+            let index = {
+                cfg_if::cfg_if! {
```
Alright then.

Hywan commented:
@jmartinesp Thanks! Can you measure if the memory peak maps to the size of the media, or is it something else?

jmartinesp commented:
<img width="1501" alt="image" src="https://github.com/user-attachments/assets/41c9a80d-bc58-4388-ad17-fc1a6b77bb1b" />

I'd say it looks quite similar without these changes, the image above being a bit more stretched.

bnjbvr commented on crates/matrix-sdk-base/src/store/send_queue.rs:
```diff
@@ -109,6 +109,22 @@ pub enum QueuedRequestKind {
+
+        /// The number of bytes in the unencrypted file (or thumbnail) to be
+        /// uploaded by this request.
+        #[serde(default)]
+        file_size: usize,
+
+        /// If this is a media file upload and a related thumbnail was
+        /// previously, uploaded, the number of bytes in the unencrypted
+        /// thumbnail.
+        #[serde(default)]
+        thumbnail_file_size: Option<usize>,
+
+        /// If this is a thumbnail upload, the number of bytes in the
+        /// unencrypted media file to be uploaded subsequently.
+        #[serde(default)]
+        media_file_size: Option<usize>,
```
I'm failing to see why any of these should be persisted: after all, this information can be inferred from the combination of the previous sent media infos (that we already have at our disposal), and using the cache store, right?

bnjbvr replied:
> For thumbnail_file_size

I'm having a hard time to follow, so please bear with me :) It looks to me that the `thumbnail_source` field should include the uploaded (so the final non-local MXC) here, or many comments are wrong in the code. Is that not the case? I would assume that, since thumbnails are uploaded before the accompanying media, and the `QueuedRequest` is pushed after the thumbnail upload completed: the source would include a non-local URL. And it should've been renamed in the media cache as well, so you could get the media from there?

> For media_file_size,

Sounds like you could load all the dependent queued requests and do the filtering outside the database. Or add a new event cache store trait method to do this kind of filtering.

---

Alternatively: since the upload progress is mostly concerned about what we do in the current user session, there might be a simpler alternative that doesn't involve the store at all: keep an in-memory mapping of (media|thumbnail) -> size. It's initially filled with some entries when creating the request, and these entries are removed when the final request (media | gallery upload) is complete.

I think that's correct too, in the case of interruptions (e.g. the process exits improperly). The next time we'll open the app, we'll be only interested about progress of the things we have *not* uploaded yet, so while the total may not include the total size for all the media/thumbnails we've already included, it should still be interesting to the user, as it shows the current progress.

Johennes replied:
The tree properties have been eliminated in the latest version of the PR. I used a mix of the media cache, dependent stored requests and an in-memory cache for thumbnail sizes.

bnjbvr commented on crates/matrix-sdk-base/src/store/send_queue.rs:
```diff
@@ -242,6 +258,22 @@ pub enum DependentQueuedRequestKind {
         #[serde(default = "default_parent_is_thumbnail_upload")]
         parent_is_thumbnail_upload: bool,
+
+        /// The number of bytes in the unencrypted file (or thumbnail) to be
+        /// uploaded by this request.
+        #[serde(default)]
+        file_size: usize,
+
+        /// If this is a media file upload and a related thumbnail was
+        /// previously, uploaded, the number of bytes in the unencrypted
+        /// thumbnail.
+        #[serde(default)]
+        thumbnail_file_size: Option<usize>,
+
+        /// If this is a thumbnail upload, the number of bytes in the
+        /// unencrypted media file to be uploaded subsequently.
+        #[serde(default)]
+        media_file_size: Option<usize>,
```
Ditto.

bnjbvr commented on crates/matrix-sdk/src/http_client/mod.rs:
```diff
@@ -231,6 +231,15 @@ pub struct TransmissionProgress {
     pub total: usize,
 }
 
+/// Progress of an operation in abstract units.
```
Can you explain in the doc comment why the units are abstract / what does it mean?

bnjbvr commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -580,7 +601,57 @@ impl RoomSendQueue {
+                let updates = updates.clone();
+                let related_to = related_to.clone();
+
+                // Watch and communicate the progress on a detached background task. Once
+                // the progress observable is dropped, next() will return None and the
+                // task will end.
+                spawn(async move {
```
Does it correctly terminate when the upload is cancelled? (needs a test!)

Johennes replied:
I moved this comment to https://github.com/matrix-org/matrix-rust-sdk/pull/5008/files#r2190753736 as this position is now outdated.

bnjbvr reviewed (changes requested):
I wonder if we could avoid storing the sizes in persistent storage; after all, we're only interested in the aggregate of the remaining files to upload, so keeping this information in memory ought to be sufficient (plus, all the previously uploaded medias should still be in the event cache store until the media event is sent, at the end of the process). Let's try to make it more stateless instead.

This PR is a bit hard to review; if you introduce renamings, please keep the commit history tidy (no merge commits, only rebase/fixup! commits) and put renamings into individual commits. Also, would be sweet to group changes to the same layer (send queue vs timeline vs FFI) in commits, so as to give me a "direction" where to start from, and where to look at, in the end. You don't *have to* do it for this very specific PR, but for what it's worth, this is the kind of empathetic changes that usually helps (thus speeds up) reviews :)

I'd like to give this more thought and another look. Also, please add at least one end-to-end test so as to validate the approach :)

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -592,9 +663,16 @@ impl RoomSendQueue {
 
                         SentRequestKey::Media(media_info) => {
-                            let _ = updates.send(RoomSendQueueUpdate::UploadedMedia {
+                            let _ = updates.send(RoomSendQueueUpdate::MediaUpload {
```
Sorry, it looks like I forgot all the good practices while iterating on this PR. 🙈

Will try to be more diligent in future.

bnjbvr commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -580,7 +601,57 @@ impl RoomSendQueue {
+
+            // Prepare to watch and communicate the request progress.
+            let progress = if let Some(related_to) = related_txn_id.as_ref() {
+                let progress: SharedObservable<TransmissionProgress> = Default::default();
```
(Ah yeah, eventually we'd get rid of `subscribe_to_send_progress` entirely, as well as the `use_send_queue` method, so the timeline/FFI will only use the send queue.)

I think it's nice to keep it as a separate field in `AttachmentConfig`. Sorry I got confused for a moment :)

> In either case, I will also have to add it on QueuedRequestKind::MediaUpload and DependentQueuedRequestKind::UploadFileOrThumbnail then to make it available here. I assume that is ok because I cannot think of another way?

We could also have a global setting, for all the `RoomSendQueue`s, that controls whether we want to observe upload progress or not. I'd imagine this would cover most use cases: either you want *all* the progress updates for medias, or you want *none* of them. I can't imagine a case where one would be interested in *some* progress updates, but not the other ones.

This way, it could be stored directly as an `AtomicBool` in `SendQueueData`, and that would be read (1) before creating the task, (2) before sending any media upload progress update (maybe this point is overkill, though?). 

What do you think?

Johennes replied:
Sounds good, have mode it so.

Johennes commented on crates/matrix-sdk-base/src/store/send_queue.rs:
```diff
@@ -379,6 +379,10 @@ pub struct SentMediaInfo {
 
+    /// The number of bytes uploaded for the file.
+    #[serde(default)]
+    pub bytes: usize,
```
This has been eliminated in the latest version of the PR.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -643,6 +643,7 @@ impl RoomSendQueue {
                             let _ = updates.send(RoomSendQueueUpdate::UploadedMedia {
                                 related_to: related_txn_id.as_ref().unwrap_or(&txn_id).clone(),
                                 file: media_info.file,
+                                bytes: media_info.bytes,
```
This has been eliminated in the latest version of the PR.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -2163,7 +2345,7 @@ pub struct SendHandle {
     /// If this is a media upload, this is the "main" transaction id, i.e. the
     /// one used to send the event, and that will be seen by observers.
-    transaction_id: OwnedTransactionId,
+    pub transaction_id: OwnedTransactionId,
```
Sorry, this was a leftover from an earlier attempt that wasn't actually needed anymore. I have removed it.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -674,13 +722,171 @@ impl RoomSendQueue {
+            let related_to = related_txn_id.clone();
+            let updates = updates.clone();
+
+            // Watch and communicate the progress on a detached background task. Once
+            // the progress observable is dropped, next() will return None and the
+            // task will end.
+            spawn(async move {
+                while let Some(progress) = subscriber.next().await {
+                    if report_media_upload_progress.load(Ordering::SeqCst) {
+                        let _ = updates.send(RoomSendQueueUpdate::MediaUpload {
+                            related_to: related_to.clone(),
+                            file: None,
+                            index: media_upload_info.index,
+                            progress: compute_combined_progress(
+                                progress,
+                                media_upload_info.bytes,
+                                media_upload_info.uploaded,
+                                media_upload_info.pending,
+                            ),
+                        });
+                    }
+                }
+            });
```
@bnjbvr [said](https://github.com/matrix-org/matrix-rust-sdk/pull/5008#discussion_r2151535506):

> Does it correctly terminate when the upload is cancelled? (needs a test!)

Johennes replied:
According to the docs, `next` will return `None` once `progress` is dropped which should break the loop and end the task. I'm struggling a little to envision an automated test for this. I can only think of sending some synthetic `RoomSendQueueUpdate::MediaUploadTaskEnded` (or similar)?

Johennes commented:
@bnjbvr this is now ready for review again. I've addressed all previous comments other than https://github.com/matrix-org/matrix-rust-sdk/pull/5008#discussion_r2190753736. Since I had previously made a mess out of the git history I also rebased everything into 10 self-contained commits. I will use only fix-ups from here on but wanted to try and make the base as easy to review as possible.

Regarding the approach, I went for a mix of using the event cache, info from dependent requests and an in-memory cache for thumbnail sizes.

Johennes commented on crates/matrix-sdk/src/encryption/futures.rs:
```diff
@@ -30,7 +30,7 @@ use crate::{config::RequestConfig, Client, Media, Result, TransmissionProgress};
     client: &'a Client,
     reader: &'a mut R,
-    send_progress: SharedObservable<TransmissionProgress>,
+    pub(crate) send_progress: SharedObservable<TransmissionProgress>,
```
Sorry, I think this is a leftover from an earlier attempt. It's not actually needed indeed. Fixed in 0f2eb69d3a4e144ac675c12a21c83a75b1a862d0.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -921,6 +1160,23 @@ struct BeingSentInfo {
+    bytes: usize,
+    /// If the current file is not a thumbnail, the total number of bytes in a
+    /// previously uploaded thumbnail, if any exists. Otherwise, zero.
+    uploaded: usize,
```
Yeah, that sounds more clear. Renamed in 48ccad930dbc00e34458ba596eda74d3c18f0f3e.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -921,6 +1160,23 @@ struct BeingSentInfo {
+    uploaded: usize,
+    /// If the current file is a thumbnail, the total number of bytes in the
+    /// related media file still to be uploaded. Otherwise, zero.
+    pending: usize,
```
Renamed in 48ccad930dbc00e34458ba596eda74d3c18f0f3e.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -701,6 +725,162 @@ impl RoomSendQueue {
+        }
+    }
+
+    /// Try to creates an observable to watch a media's upload progress.
```
Oops. Fixed in 13631d545bf0165fa384bda5c32f8024310d6a47.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -701,6 +725,162 @@ impl RoomSendQueue {
+            // task will end.
+            spawn(async move {
+                while let Some(progress) = subscriber.next().await {
+                    if report_media_upload_progress.load(Ordering::SeqCst) {
```
Yeah, you're right. It's unlikely that somebody would care about this taking effect for an already ongoing upload. Have removed the check inside the spawned task with 50696e1fe299da30fc591b10cc71cb02178f6749.

Johennes commented on crates/matrix-sdk-ui/src/timeline/event_item/local.rs:
```diff
@@ -84,3 +87,19 @@ pub enum EventSendState {
+        index: u64,
+
+        /// The combined upload progress across the file and, if existing, its
+        /// thumbnail.
```
I happened to already fix this when addressing https://github.com/matrix-org/matrix-rust-sdk/pull/5008#discussion_r2194738467.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -851,6 +1031,65 @@ impl RoomSendQueue {
+        current: uploaded + progress.current,
+        total: uploaded + progress.total + pending,
+    }
+}
```
Good idea. Fixed in 0e1baace4ac32457b8a67bbf0f9182661264fd58.

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -964,6 +967,10 @@ struct QueueStorage {
+
+    /// In-memory mapping of media transaction IDs to thumbnail sizes for the
+    /// purpose of progress reporting.
+    thumbnail_size_cache: Arc<Mutex<HashMap<OwnedTransactionId, Vec<Option<usize>>>>>,
```
I addressed 3 with 971dcdafb5586dd001906444cfe529ce7415db6c and ad963c786d8ac590296d9711b7719d7f506cabc0. I had to switch back to the async `Mutex` though because that has `lock_owned` and I couldn't figure out the lifetimes without it. 🙈

Johennes commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -701,6 +725,162 @@ impl RoomSendQueue {
+
+    /// Determine the size of a pending file upload, if this is a thumbnail
+    /// upload or return 0 otherwise.
+    async fn get_dependent_pending_file_upload_size(
```
I could also store the other needed sizes there, yes. I think we would still need the current look-ups from persisted data though because in the case of galleries, we would otherwise not have anything to report progress for the remaining gallery items?

Johennes commented:
> Usually at this point, I'd try to help with the merging of this PR, by addressing the comments myself, but turns out I'll be away starting today and for a week or so, so I don't have much time to do this. If you're bored with this PR, I'll be happy to address my own comments when I'm back. Alternatively, another reviewer can make sure that my comments have been addressed in the meanwhile (although it'd be a bit of a stretch for them to read all the prior work).

Thanks for the review. 🙏 Letting this sit until you return seems fine from my perspective. I have addressed all comments apart from those I left unresolved.

bnjbvr commented on crates/matrix-sdk-ui/src/timeline/event_item/local.rs:
```diff
@@ -65,7 +65,10 @@ impl LocalEventTimelineItem {
 pub enum EventSendState {
     /// The local event has not been sent yet.
-    NotSentYet,
+    NotSentYet {
+        /// The progress of the sending operation, if any is available.
+        progress: Option<EventSendProgress>,
+    },
```
Based on other comments: we might prefer to have a new variant `NotUploadedYet`, for medias. The `progress` field isn't used, for non-media uploads.

bnjbvr commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -724,13 +769,180 @@ impl RoomSendQueue {
+            return None;
+        }
+
+        if let Some(related_txn_id) = related_txn_id {
```
let's avoid one indent level, since this function returns an `Option` :)

`let related_txn_id = related_txn_id?;`

bnjbvr commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -618,7 +638,23 @@ impl RoomSendQueue {
 
-            match Self::handle_request(&room, queued_request, cancel_upload_rx).await {
+            // Prepare to watch and communicate the request progress for media uploads.
+            let media_upload_info =
+                RoomSendQueue::try_create_media_upload_info(&queued_request, &room, &queue)
+                    .await
+                    .unwrap_or_default();
+            let progress = RoomSendQueue::try_create_media_upload_progress_observable(
+                &report_media_upload_progress,
+                &media_upload_info,
+                &related_txn_id,
+                room_id,
+                &global_update_sender,
+                &update_sender,
+            );
```
This code will always create a progress listener, even for sending non-medias. I don't think we should do this, so maybe it is time to split the code in such a way that sending a plain event or a media go through different code paths. Let's not act on it quite yet; i'd like to think more about this.

bnjbvr commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -724,13 +769,180 @@ impl RoomSendQueue {
+
+        // Get the size of the file being uploaded from the event cache.
+        let bytes = if let Ok(cache) = room.client().event_cache_store().lock().await {
+            if let Ok(Some(content)) = cache.get_media_content(cache_key).await {
```
We really can't just ignore errors, as they're unexpected. We need proper error handling in this function.

bnjbvr commented on crates/matrix-sdk/src/send_queue/mod.rs:
```diff
@@ -724,13 +769,180 @@ impl RoomSendQueue {
+        }
+
+        if let Ok(cache) = room.client().event_cache_store().lock().await {
+            if let Ok(Some(content)) = cache.get_media_content(&cache_key).await {
```
We need proper error handling here too.

bnjbvr reviewed (changes requested):
Sorry, it seems that taking a week of vacation gave me a fresh pair of eyes, and I spotted many new surprising things I hadn't seen before.

I'd like this PR to be split between the timeline changes and the send queue changes, otherwise we're likely to remain stuck in review land for too long, as there are still many changes to address at both layers.

I've posted a few more comments below, but does it sound reasonable to break the PR apart in smaller chunks again, please? One thing I could do, as a proof of goodwill, is try to hack on the send queue related changes myself, by making PRs against a branch that only included the send queue related changes.

Johennes commented on bindings/matrix-sdk-ffi/src/client.rs:
```diff
@@ -224,6 +224,30 @@ impl From<matrix_sdk::TransmissionProgress> for TransmissionProgress {
     }
 }
 
+/// Progress of an operation in abstract units.
+///
+/// Contrary to [`TransmissionProgress`], this allows tracking the progress
+/// of sending or receiving a payload in estimated pseudo units representing a
+/// percentage. This is helpful in cases where the exact progress in bytes isn't
+/// known, for instance, because encryption (which changes the size) happens on
+/// the fly.
+#[derive(Clone, Copy, uniffi::Record)]
+pub struct AbstractProgress {
+    /// How many units were already transferred.
+    pub current: u64,
+    /// How many units there are in total.
+    pub total: u64,
+}
+
+impl From<matrix_sdk::AbstractProgress> for AbstractProgress {
+    fn from(value: matrix_sdk::AbstractProgress) -> Self {
+        Self {
+            current: value.current.try_into().unwrap_or(u64::MAX),
+            total: value.total.try_into().unwrap_or(u64::MAX),
+        }
+    }
+}
+
```
Fixed in 5ec82b101e9a11885ff4518a4f0de6168edba93c.

Johennes commented on crates/matrix-sdk-ui/tests/integration/timeline/media.rs:
```diff
@@ -141,14 +142,36 @@ async fn test_send_attachment_from_file() {
 
+    // The media upload finishes.
+    {
+        assert_let_timeout!(
+            Duration::from_secs(3),
+            Some(VectorDiff::Set { index: 1, value: item }) = timeline_stream.next()
+        );
```
Yes, the flag only controls the sending of incremental updates during upload. The update signalling upload completion is always sent because this logic already existed before.

Johennes commented on crates/matrix-sdk-ui/tests/integration/timeline/media.rs:
```diff
@@ -141,14 +142,36 @@ async fn test_send_attachment_from_file() {
         );
         assert_let!(Some(msg) = item.content().as_message());
-        assert_matches!(item.send_state(), Some(EventSendState::NotSentYet));
+        assert_matches!(item.send_state(), Some(EventSendState::NotSentYet { progress: None }));
```
Fixed in c23d92d9f6a30d974b61a6e1ad1167842f7f1402.
